File: ccc/cpp/ccc13s4.cpp
#include <cstdio>
#include <cstring>
#include <queue>
#include <stack>
#include <vector>

std::vector<std::vector<int>> graph;

bool bfs(int start, int end) {
    bool *visited = new bool[graph.size()]();
    bool *queued = new bool[graph.size()]();
    std::queue<int> q;
    q.push(start);

    while(!q.empty()) {
        auto n = q.front();
        q.pop();
        visited[n] = true;

        if(n == end) {
            delete[] visited;
            delete[] queued;
            return true;
        }

        for(const int &child : graph[n]) {
            if(!visited[child] && !queued[child]) {
                queued[child] = true;
                q.push(child);
            }
        }
    }
    delete[] visited;
    delete[] queued;
    return false;
}

int main() {
    int n, m;
    std::scanf("%d%d", &n, &m);

    graph.resize(n);

    for(int i = 0; i < m; i ++) {
        int a, b;
        std::scanf("%d%d", &a, &b);

        graph[a - 1].push_back(b - 1);
    }

    int p, q;
    std::scanf("%d%d", &p, &q);
    p -= 1;
    q -= 1;

    bool taller = bfs(p, q);
    bool shorter = bfs(q, p);
    if(taller) {
        std::printf("yes\n");
    }
    else if(shorter) {
        std::printf("no\n");
    }
    else {
        std::printf("unknown\n");
    }
}

File: ccc/py/ccc13s4.py
"""
Python solution for CCC '13 J4 - Who is Taller?
"""

import sys
_all_data = sys.stdin.read().split('\n')
_line = -1
def _input():
    global _line
    _line += 1
    return _all_data[_line]
input = _input

from queue import Queue
n, m = input().split()
n = int(n)
m = int(m)

graph = [[] for i in range(n)]

for i in range(m):
    people = input().split()
    graph[int(people[0]) - 1].append(int(people[1]) - 1)

p, q = input().split()
p = int(p) - 1
q = int(q) - 1

def bfs(graph, start, end):
    visited = [False for i in range(len(graph))]
    queued = [False for i in range(len(graph))]
    q = Queue()
    q.put(start)

    while not q.empty():
        n = q.get()
        visited[n] = True

        if n == end:
            return True

        for child in graph[n]:
            if not visited[child] and not queued[child]:
                queued[child] = True
                q.put(child)
    return visited[end]

taller = bfs(graph, p, q)
shorter = bfs(graph, q, p)

print("yes" if taller else "no" if shorter else "unknown")
